llvm/lib/CheerpUtils/PartialExecuter.cpp

```diff
@@ -0,0 +1,1393 @@
+	//TODO(carlo): an optmization might be having from be a set<BasicBlock>, conserving the phi that are equals
+	BasicBlockGroupData* parent;
+	std::deque<BasicBlockGroupData> subGroups;
+	std::map<const llvm::BasicBlock*, int> subGroupsID;
+	bool visitingAll{false};
+	int iterationCount{-1};
+	std::vector<std::pair<const llvm::BasicBlock*, int> > escaping;
+	static const DeterministicBBSet getAllBasicBlocks(const llvm::Function& F)
+	{
+		DeterministicBBSet set;
+		for (const llvm::BasicBlock& bb : F)
+		{
+			set.insert(&bb);
+		}
+		return set;
+	}
+	void splitIntoSCCs(std::deque<BasicBlockGroupData>& blockQueue, std::map<const llvm::BasicBlock*, int>& blockToIndexMap);
+public:
+	BasicBlockGroupData(FunctionData& data, BasicBlockGroupData* parentBBGD, const DeterministicBBSet& inner, const llvm::BasicBlock* start = nullptr)
+		: data(data), inner(inner), start(start), from(nullptr), parent(parentBBGD)
+	{
+		if (start)
+			assert(start->getParent() == data.getFunction());
+	}
+	BasicBlockGroupData(FunctionData& data)
+		: BasicBlockGroupData(data, /*parentBBGD*/nullptr, getAllBasicBlocks(*data.getFunction()), &data.getFunction()->getEntryBlock())
+	{
+	}
+	BasicBlockGroupData(BasicBlockGroupData& BBGData)
+		: data(BBGData.data), inner(BBGData.inner), start(nullptr), from(nullptr), parent(&BBGData)
+	{
+	}
+	void addIncomingEdge(const llvm::BasicBlock* comingFrom, const llvm::BasicBlock* target)
+	{
+		assert(inner.count(target));
+		if (start == nullptr)
+		{
+			start = target;
+			from = comingFrom;
+		}
+		if (start != target)
+		{
+			multiHead = true;
+			from = nullptr;
+		}
+		if (comingFrom != from)
+		{
+			from = nullptr;
+		}
+	}
+	void runVisitBasicBlock(const llvm::BasicBlock& BB, std::vector<const llvm::BasicBlock*>& visitNext)
+	{
+		assert(visitNext.empty());
+
+		data.currentEE->setIncomingBB(from);
+		const BasicBlock* ret = data.currentEE->visitBasicBlock<PartialInterpreter::VisitingPolicy::COMPUTE_VALUES>(&BB, from);
+
+		if (ret)
+		{
+			visitNext.push_back(ret);
+		}
+		else
+		{
+			llvm::DenseSet<const llvm::BasicBlock*> setSuccessors;
+			for (auto* bb : successors(&BB))
+			{
+				if (setSuccessors.insert(bb).second)
+					visitNext.push_back(bb);
+			}
+		}
+		//Do the visit of the BB, with comingFrom (possibly nullptr) as predecessor
+		//Loop backs will be directed to another BBgroup
+		//The visit will return the set of reachable BBs -> add them to the subGroups data accordingly
+	}
+	bool shouldCleanUp()
+	{
+		for (auto& p : escaping)
+		{
+			//llvm::errs() <<"escaping:\t" << p.first->getName() << "\t" << p.second <<"\n";
+			if (data.visitCounter[p.first] != p.second)
+				return true;
+		}
+		return false;
+	}
+	bool notifySuccessor(const llvm::BasicBlock* from, const llvm::BasicBlock* succ, const int iter)
+	{
+		if (inner.count(succ) == 0)
+		{
+			assert(parent);
+			//It should handled by the parent SCC
+
+			const bool notified = parent->notifySuccessor(from, succ, iter);
```

Comment:
Unsure about this recursive calling, what is the idea?

Reply:
I will add explainer. Idea is there is that this is basically a tree, with the root containing everything, and going down there are basically subsets of the BBs. If we start at a given depth, and we loop for node succ, and it's not part of the current set of BB, then has to be up in the hierarchy.
I will write down this properly (setting as resolved when I have done)

Reply:
Added explainer